src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -0,0 +1,153 @@
+     *  - the buyer pays [[leaseFeeProportional]] (expressed in basis points) of the amount contributed by the seller
+     *  - the buyer refunds the on-chain fees for up to [[fundingWeight]] of the utxos contributed by the seller
+     *
+     * The seller promises that their relay fees towards the buyer will never exceed [[maxRelayFeeBase]] and [[maxRelayFeeProportional]].
+     * This cannot be enforced, but if the buyer notices the seller cheating, they should blacklist them and can prove
+     * that they misbehaved.
+     */
+    data class LeaseRates(val fundingWeight: Int, val leaseFeeProportional: Int, val maxRelayFeeProportional: Int, val leaseFeeBase: Satoshi, val maxRelayFeeBase: MilliSatoshi) {
+        val maxRelayFeeProportionalMillionths: Long = maxRelayFeeProportional.toLong() * 100
+
+        /**
+         * Fees paid by the liquidity buyer: the resulting amount must be added to the seller's output in the corresponding
+         * commitment transaction.
+         */
+        fun fees(feerate: FeeratePerKw, requestedAmount: Satoshi, contributedAmount: Satoshi): Satoshi {
+            val onChainFees = Transactions.weight2fee(feerate, fundingWeight)
+            // If the seller adds more liquidity than requested, the buyer doesn't pay for that extra liquidity.
+            val proportionalFee = requestedAmount.min(contributedAmount) * leaseFeeProportional / 10_000
+            return leaseFeeBase + proportionalFee + onChainFees
+        }
+
+        fun signLease(nodeKey: PrivateKey, localFundingPubKey: PublicKey, requestFunds: ChannelTlv.RequestFunds): ChannelTlv.WillFund {
+            val witness = LeaseWitness(localFundingPubKey, requestFunds.leaseExpiry, requestFunds.leaseDuration, maxRelayFeeProportional, maxRelayFeeBase)
+            val sig = witness.sign(nodeKey)
+            return ChannelTlv.WillFund(sig, this)
+        }
+
+        fun write(out: Output) {
+            LightningCodecs.writeU16(fundingWeight, out)
+            LightningCodecs.writeU16(leaseFeeProportional, out)
+            LightningCodecs.writeU16(maxRelayFeeProportional, out)
+            LightningCodecs.writeU32(leaseFeeBase.sat.toInt(), out)
+            LightningCodecs.writeTU32(maxRelayFeeBase.msat.toInt(), out)
+        }
+
+        companion object {
+            fun read(input: Input): LeaseRates = LeaseRates(
+                fundingWeight = LightningCodecs.u16(input),
+                leaseFeeProportional = LightningCodecs.u16(input),
+                maxRelayFeeProportional = LightningCodecs.u16(input),
+                leaseFeeBase = LightningCodecs.u32(input).sat,
+                maxRelayFeeBase = LightningCodecs.tu32(input).msat,
+            )
+        }
+    }
+
+    /** Request inbound liquidity from a remote peer that supports liquidity ads. */
+    data class RequestRemoteFunding(val fundingAmount: Satoshi, val maxFee: Satoshi, val leaseStart: Int, val leaseDuration: Int) {
```

Comment:
I think it would make more sense to provide the actual liquidity parameters here, (that we would get either from conf, our direct peer, network announcements...).

```suggestion
    data class RequestRemoteFunding(val fundingAmount: Satoshi, val leaseRates: LeaseRates) {
```

Reply:
Done in https://github.com/ACINQ/lightning-kmp/pull/561/commits/c65c4cf919c3572095bc997868e8a01dbabb8ede